File: 88-merge-sorted-array/merge-sorted-array.ts
/**
 Do not return anything, modify nums1 in-place instead.
 */
function merge(nums1: number[], m: number, nums2: number[], n: number): void {
    let first = m - 1 , second = n - 1 , end = m + n - 1;

    while(first >=0 && second >= 0){
        if(nums1[first] >= nums2[second]){
            let a = nums1[end];
            let b = nums1[first];

            nums1[first] = a;
            nums1[end] = b;

            end--,first--;
        }
        else{
            nums1[end--] = nums2[second--];
        }
    }

    //if there is something else remained in second array then copy them all in second
    while(second >=0 ){
        nums1[end--] = nums2[second--];
    }
};

File: 88-merge-sorted-array/merge-sorted-array.cpp
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int first = m -1, second = n -1, end= m+n-1;

        if(n==0) return;

        if(m == 0){
            while(second >=0){
                nums1[end--] = nums2[second--];
            }
        }

        while(first >=0 && second >= 0){
            if(nums1[first] >= nums2[second]){
                swap(nums1[first--], nums1[end]);
            } else{
                nums1[end] = nums2[second--];
            }

            end--;
        }

        while(second >= 0){
              nums1[end--] = nums2[second--];
        }
    }
};